packages/metrics/src/Metrics.ts

```diff
@@ -692,28 +709,51 @@ class Metrics extends Utility implements MetricsInterface {
       {}
     );
 
-    const dimensionNames = [
-      ...new Set([
+    const dimensionNames = [];
+
+    const allDimensionKeys = new Set([
+      ...Object.keys(this.defaultDimensions),
+      ...Object.keys(this.dimensions),
+    ]);
+
+    if (Object.keys(this.dimensions).length > 0) {
+      dimensionNames.push([...allDimensionKeys]);
+    }
+
+    for (const dimensionSet of this.dimensionSets) {
+      const dimensionSetKeys = new Set([
         ...Object.keys(this.defaultDimensions),
-        ...Object.keys(this.dimensions),
-      ]),
-    ];
+        ...Object.keys(dimensionSet),
+      ]);
+      dimensionNames.push([...dimensionSetKeys]);
+    }
+
+    if (
+      dimensionNames.length === 0 &&
+      Object.keys(this.defaultDimensions).length > 0
+    ) {
+      dimensionNames.push([...Object.keys(this.defaultDimensions)]);
+    }
 
     return {
       _aws: {
         Timestamp: this.#timestamp ?? new Date().getTime(),
         CloudWatchMetrics: [
           {
             Namespace: this.namespace || DEFAULT_NAMESPACE,
-            Dimensions: [dimensionNames],
+            Dimensions: dimensionNames as [string[]],
             Metrics: metricDefinitions,
           },
         ],
       },
-      ...this.defaultDimensions,
-      ...this.dimensions,
-      ...metricValues,
-      ...this.metadata,
+      ...Object.assign(
```

Comment:
Can we put back the spread operator (`...`)? With current implementation we're instantiating a new object (with `Object.assing()`) only to immediately spread it.

Reply:
Done